[git-webkit] Provide mechanism to exempt bugs from redaction

JonWBedard opened the pull request:
#### e7a1365dc511795c1924d5248d10397aacf02604
<pre>
[git-webkit] Provide mechanism to exempt bugs from redaction
<a href="https://bugs.webkit.org/show_bug.cgi?id=254993">https://bugs.webkit.org/show_bug.cgi?id=254993</a>
rdar://107615755

Reviewed by Geoffrey Garen.

It is often the case that we with to (eventually) publish changes which are
redacted when they are written. git-webkit should support a mechanism to mark
bugs (and by extension, the commits associated with those bug) as exempt from
redaction.

* Tools/Scripts/hooks/pre-push: Do not flag class 3 commits if an issue is
exempt from redaction.
* Tools/Scripts/libraries/webkitbugspy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py:
(Tracker.__init__): Pass redact_exemption.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/github.py:
(Tracker.__init__): Pass redact_exemption.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/issue.py:
(Issue.redacted): Check if an issue is exempt from redaction.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/radar.py:
(Tracker.__init__): Pass redact_exemption.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tests/bugzilla_unittest.py:
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tests/github_unittest.py:
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tests/radar_unittest.py:
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tracker.py:
(Tracker.Redaction): Support defining a redaction exemption.
(Tracker.from_json): Pass redact_exemption.
(Tracker.__init__): Ditto.
* Tools/Scripts/libraries/webkitscmpy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull_request.py:
(PullRequest.create_pull_request):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/pull_request_unittest.py:
* metadata/trackers.json: Exempt radars with the &apos;WebKit Cleared for Publication&apos; keyword.

Canonical link: <a href="https://commits.webkit.org/262617@main">https://commits.webkit.org/262617@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c9608ea7ac23d6c9e24b89c92e70c4e890b53773

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/1983 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/2012 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/2073 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/2907 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/2058 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/1956 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/2088 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/2053 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/1820 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/2000 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/1778 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/1776 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/2754 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/2014 "Passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/1768 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/1758 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/1667 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/1819 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/1805 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/2923 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/1826 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/1627 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/1968 "Passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/1752 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/1769 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/1924 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/225 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tracker.py:
```diff
@@ -55,6 +56,10 @@ def __nonzero__(self):
+            if self.exemption:
+                if self.reason:
+                    return '{} and is exempt from redaction'.format(self.reason)
+                return 'is exempt from redaction for an unknown reason'
```
Why do we want to support the "unknown reason" case? Seems like that should be a hard error.

geoffreygaren replied:
One potential way to implement this suggestion is to change exemption from a boolean with a separate reason field to a combined exemptionReason field, which is either a string or null. (Or maybe use an enum instead.)

JonWBedard replied:
We don't really support the "unknown reason" case, I was just being thorough in exhausting cases. I'll raise an exception in the constructor, in practice, this should be generated from `metadata/trackers.json` where we define reason in all cases.

geoffreygaren commented on metadata/trackers.json:
```diff
@@ -23,6 +23,7 @@
             "tentpole:.*Security.*": true
-        }
+        },
+        "redact_exemption" : {"keywords:[^:]*WebKit Cleared for Publication" : true}
```
If we can tie this keyword to the condition it hopes to express, then we can reduce the chances that someone will apply it blindly just to merge their patch. How about:

WebKit Security Does Not Repro in Shipping

Is there any condition where a Security bug does repro in shipping, but we want to support publishing it anyway?

JonWBedard replied:
We would need to use this same mechanism after shipping security fixes too during merge-back, because all merge-back commits will be referencing security commits which did reproduce in the latest shipping release when the bug was first filed.

We can also support multiple keywords here, so could have multiple supported keywords.

geoffreygaren reviewed (approved):
r=me with some suggestions

JonWBedard committed:
[git-webkit] Provide mechanism to exempt bugs from redaction
https://bugs.webkit.org/show_bug.cgi?id=254993
rdar://107615755

Reviewed by Geoffrey Garen.

It is often the case that we with to (eventually) publish changes which are
redacted when they are written. git-webkit should support a mechanism to mark
bugs (and by extension, the commits associated with those bug) as exempt from
redaction.

* Tools/Scripts/hooks/pre-push: Do not flag class 3 commits if an issue is
exempt from redaction.
* Tools/Scripts/libraries/webkitbugspy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py:
(Tracker.__init__): Pass redact_exemption.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/github.py:
(Tracker.__init__): Pass redact_exemption.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/issue.py:
(Issue.redacted): Check if an issue is exempt from redaction.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/radar.py:
(Tracker.__init__): Pass redact_exemption.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tests/bugzilla_unittest.py:
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tests/github_unittest.py:
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tests/radar_unittest.py:
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tracker.py:
(Tracker.Redaction): Support defining a redaction exemption.
(Tracker.from_json): Pass redact_exemption.
(Tracker.__init__): Ditto.
* Tools/Scripts/libraries/webkitscmpy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull_request.py:
(PullRequest.create_pull_request):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/pull_request_unittest.py:
* metadata/trackers.json: Exempt radars with the 'WebKit Cleared for Publication' keyword.

Canonical link: https://commits.webkit.org/262617@main
